ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/SourceBuilderUtils.java

```diff
@@ -83,14 +84,20 @@ static LogicalSchema buildSchema(
         .withPseudoAndKeyColsInValue(windowed, source.getPseudoColumnVersion());
   }
 
+  static FormatInfo getFormatInfo(final FormatInfo formatInfo, final QueryContext queryContext) {
+    return queryContext.getContext().contains(MATERIALIZE_OP_NAME)
+        ? formatInfo.copyWithoutProperty(ConnectProperties.SCHEMA_ID)
+        : formatInfo;
+  }
+
   static Serde<GenericRow> getValueSerde(
       final RuntimeBuildContext buildContext,
       final SourceStep<?> streamSource,
       final PhysicalSchema physicalSchema,
       final QueryContext queryContext
   ) {
     return buildContext.buildValueSerde(
-        streamSource.getFormats().getValueFormat(),
+        getFormatInfo(streamSource.getFormats().getValueFormat(), queryContext),
```

Comment:
Hey @bvarghese1 I'm confused about this change. This `getValueSerde()` helper method is being used everywhere in SourceBuilder, including where we specify the serdes to be used to deserialize from the source topic. If we strip SCHEMA_ID from there, doesn't that mean the deserializing isn't taking SCHEMA_ID into consideration any more?

Either my understanding is wrong, or this change breaks the SCHEMA_ID feature, or something else is going on that I'm missing. If you could help reconcile my confusion, that'd be much appreciated.

Reply:
The SCHEMA_ID is removed only if the QueryContext contains MATERIALIZED. Eg: if the the QueryContext contains "Left/Source" the SCHEMA_ID is not removed because we want to honor the user specified SCHEMA_ID.

Reply:
I see... I don't think we should overload this method like this, for a number of reasons. For one, I would not expect this method to modify formats from the method name, so it makes reading the code much less intuitive. It's also not guaranteed that the only time we will need to remove SCHEMA_ID (from SourceBuilder) is when materializing. That happens to be true today, but it's very easy for someone to modify SourceBuilder and blindly call this method without realizing that it needs to be modified. 

Instead of conditionally removing SCHEMA_ID from this method (which is called everywhere from SourceBuilder), can you just remove it (and call it explicitly) from the method which builds serdes for the materialized instance instead?

Reply:
Good point.